### ISSUE TITLE
Client::query with placeholders

### ISSUE THREAD

**Issue by CAFxX** [brianmario/mysql2#885]
Client::query has no way to perform a call with placeholders (e.g. `SELECT name FROM t WHERE id = ?`) like Client::prepare/Statement::execute do.

Ideally `Client::query(query, *args)` should behave like `Client::prepare(query).execute(*args)`

**Comment by sodabrew** [comment]
This is why prepared statements exist.

**Comment by CAFxX** [comment]
Sorry, I don't see how the fact that the client API fails to provide such a feature matter in this case. If a feature is missing, a wrapper (like this gem) should be able to happily implement it.

**Comment by brianmario** [comment]
@CAFxX hi! 👋 

> AFAICT they exist mainly to avoid the overhead of parsing the same query multiple times.

So, yes - this is _also_ true of prepared statements. But, the query argument replacement is a primary use-case of them as well. In fact some folks use prepared statements solely for that purpose! Also note that last time I checked, MySQL doesn't actually do any pre-optimization of queries via the prepared statement API :\ It may pre-parse the query _string_ itself, but we didn't see any pre-optimization happening. That may be different now with newer versions...

> Also prepared statements can't be safely used with :reconnect so they are not really viable alternatives.

For future reference, the `reconnect` behavior of libmysqlclient is considered harmful - in my opinion at least. All connection-state is reset, including any open transactions. Check out the [documentation for libmysqlclient](https://dev.mysql.com/doc/refman/5.5/en/c-api-auto-reconnect.html) for more detail on that:

```
The connection-related state is affected as follows:

* Any active transactions are rolled back and autocommit mode is reset.
* All table locks are released.
* All TEMPORARY tables are closed (and dropped).
* Session system variables are reinitialized to the values of the corresponding global system variables, including system variables that are set implicitly by statements such as SET NAMES.
* User variable settings are lost.
* Prepared statements are released.
* HANDLER variables are closed.
* The value of LAST_INSERT_ID() is reset to 0.
* Locks acquired with GET_LOCK() are released.
```

>> The MySQL client API does not provide a one-off parsed query interface.
>
> Sorry, I don't see how this is relevant to the discussion. If a feature is missing, a wrapper (like this gem) should be able to easily implement it.

I would argue that this driver is just that, a driver. Intentionally low-level. As a result we favor performance over almost all else. If a user of the driver wants to use argument replacement functionality, the prepared statements API was designed for exactly that purpose. But, if you want something Ruby-based there are plenty of higher-level libraries designed to do this sort of auto-escaping for you like ActiveRecord, Sequel, and DataMapper. I realize adding the overhead of an ORM just for query-string-argument-replacement-like functionality may feel like a bit much, but I'd prefer to keep mysql2 a lower-level driver on top of which functionality like this could be added.

That said, while I personally don't have time to play with something like this (not sure about @sodabrew, but I assume he's busy as well). But if you had some time to implement this functionality (in C preferably), with benchmarks showing it doesn't add much overhead - we could continue this discussion there.

**Comment by sodabrew** [comment]
In regards to adding parameter replacements on the client-side, the major issue is that you must _parse_ the query in order to identify the replacement parameters. Such parsing has to be substantially similar to MySQL's parsing [*], but without benefit of reusing the code directly, because it otherwise lives only on the server side, and with concern for the round-trip of conversions that would take place.

Consider this trivial example: `INSERT INTO foo (colA, colB) VALUES (?, ?)`. It's easy to just replace the question marks with the values from arguments. It's also easy to correctly stringify [*] and escape the arguments (i.e. with `Mysql2::Client#escape`) as needed.

But now consider this example: `INSERT INTO foo (colA, colB) VALUES ('Hello', 'World?')`. In order to determine that this question mark is quoted, one needs a parser, not just  a string replacer. A reasonable argument could be made that the parser only needs to count for escaping rules and not for complete SQL grammar.

But then consider this example: `INSERT INTO foo (?, colB) VALUES (??)`. Without a full parser, the client-side would fail to reject this query for its errors, or by accepting this query, we've created a new micro-dialect of SQL that only the gem understands. (A bound parameter cannot be in the place of an SQL keyword, table name, or field name, and two question marks is never valid in MySQL - but is valid in PostgreSQL in some cases - so you can imagine that feature might be borrowed into MySQL in the future - or might already be valid in a MySQL-protocol-alike, such as Amazon Aurora or Sphinx Search).

The two [*] notes above call out security concerns: if the two parsers are different, there's a possibility that your downstream code may appear sane, but hides a SQL injection vector if untrusted input is passed into MySQL without proper escaping. Prepared Statements are often used exactly because they avoid SQL injection by separating the query string from untrusted user input, which gets passed to the server using a completely different underlying mechanism in the protocol, and is not converted to/from a character representation.

In short, this is not a feature I would consider pursuing in this gem, which as @brianmario points out, is much lower-level than an ORM or other SQL abstraction tool that does implement such features, for example Sequel: http://sequel.jeremyevans.net/
